dataset/src/main/scala/frameless/TypedDataset.scala

```diff
@@ -106,15 +105,22 @@ class TypedDataset[T] protected[frameless](val dataset: Dataset[T])(implicit val
         i3: TypedEncoder[Out]
       ): TypedDataset[Out] = {
 
-        val cols = columns.toList[UntypedExpression[T]].map(c => new Column(c.expr))
-
-        val selected = dataset.toDF()
-          .agg(cols.head.alias("_1"), cols.tail: _*)
-          .as[Out](TypedExpressionEncoder[Out])
-          .filter("_1 is not null") // otherwise spark produces List(null) for empty datasets
-
-        TypedDataset.create[Out](selected)
-      }
+      val underlyingColumns = columns.toList[UntypedExpression[T]]
+      val cols: Seq[Column] = for {
+        (c, i) <- columns.toList[UntypedExpression[T]].zipWithIndex
+      } yield new Column(c.expr).as(s"_${i+1}")
+
+      // Workaround to SPARK-20346. The alternative is to allow the result to be Vector(null) for empty DataFrames.
+      val filterStr = (
+        for {
+          (c, i) <- underlyingColumns.zipWithIndex
+          if !c.uencoder.nullable
+        } yield s"_${i+1} is not null"
+        ).mkString(" or ")
+
+      val selected = dataset.toDF().agg(cols.head, cols.tail:_*).as[Out](TypedExpressionEncoder[Out])
+      TypedDataset.create[Out](if (filterStr.isEmpty) selected else selected.filter(filterStr))
```

Comment:
My only concern here is that this might be expensive. It might be good to benchmark that vs the default behavior to see how much we lose.

Reply:
The result of agg would be a single row storing the aggregated results. So this filter only applies to a single row.